File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        HashMap<Character, Integer> countChar = new HashMap<>();
        int left = 0, maxLen = 0, maxFreq = 0;

        for (int right = 0; right < s.length(); right++) {
            // Count the frequency of the current character
            countChar.put(s.charAt(right), countChar.getOrDefault(s.charAt(right), 0) + 1);
            // Track the frequency of the most frequent character in the window
            maxFreq = Math.max(maxFreq, countChar.get(s.charAt(right)));

            // If the current window size minus the most frequent character's count is greater than k, shrink the window
            if ((right - left + 1) - maxFreq > k) {
                countChar.put(s.charAt(left), countChar.get(s.charAt(left)) - 1);
                left++;
            }

            // Update the maximum length of the valid window
            maxLen = Math.max(maxLen, right - left + 1);
        }

        return maxLen;
    }
}

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        countChar = [0]*26
        left = 0
        right = 0
        maxLen = 0
        maxFreq = 0
        
        while right < len(s):
            countChar[ord(s[right]) - ord('A')] += 1
            maxFreq = max(maxFreq, countChar[ord(s[right]) - ord('A')])
            
            if((right - left + 1) - maxFreq > k): # the no of replace yemideregu char
                    countChar[ord(s[left]) - ord('A')] -= 1
                    left += 1
            
            maxLen = max(maxLen, right - left + 1)
            right += 1
        
        return maxLen
        